File: LupusaBlazorDemos/LBDJsInterop.cs
using LupusaBlazorDemos.Demos;
using Microsoft.JSInterop;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace LupusaBlazorDemos
{
    public static class LBDJsInterop
    {

        public static IJSRuntime jsRuntime;

        public static ValueTask<int> CalcFib(IJSRuntime jsRuntime, int num)
        {
            return jsRuntime.InvokeAsync<int>(
                "LBDJsFunctions.calcfib",
                num);
        }

        public static ValueTask<string> GenerateNewUser(IJSRuntime jsRuntime)
        {
            return jsRuntime.InvokeAsync<string>("LBDJsFunctions.generateNewUser");
        }

        internal static ValueTask<string> Alert(string message)
        {
            return jsRuntime.InvokeAsync<string>(
                "LBDJsFunctions.Alert", message);
        }

        internal static void ProcessData(string variableName)
        {
            jsRuntime.InvokeVoidAsync("LBDJsFunctions.ProcessData", variableName);
        }


        internal async static ValueTask<bool> HasFile(string inputFileElementID)
        {
            return await jsRuntime.InvokeAsync<bool>("LBDJsFunctions.HasFile", inputFileElementID);

        }

        internal async static ValueTask<bool> ReadFile(string variableName, string inputFileElementID)
        {
            return await jsRuntime.InvokeAsync<bool>("LBDJsFunctions.ReadFile", variableName, inputFileElementID);

        }

        internal async static ValueTask<string> GetFile(string variableName, string inputFileElementID)
        {
            return await jsRuntime.InvokeAsync<string>("LBDJsFunctions.GetFile", variableName, inputFileElementID);

        }

        internal static ValueTask<bool> SetData(string variableName, string t)
        {
            return jsRuntime.InvokeAsync<bool>("LBDJsFunctions.SetData", variableName, t);
        }

        internal static ValueTask<string> GetData(string variableName)
        {
            return jsRuntime.InvokeAsync<string>("LBDJsFunctions.GetData", variableName);
        }

        internal static ValueTask<bool> HandleDrag(IJSRuntime jsRuntime, string elementID, int id, DotNetObjectReference<PageDragAndDrop> dotnetHelper)
        {
            return jsRuntime.InvokeAsync<bool>(
                "LBDJsFunctions.handleDragStart", elementID, id, dotnetHelper);
        }


        internal static ValueTask<bool> HandleDrop(IJSRuntime jsRuntime, string elementID, int id, DotNetObjectReference<PageDragAndDrop> dotnetHelper)
        {
            return jsRuntime.InvokeAsync<bool>(
                "LBDJsFunctions.handleDrop", elementID, id, dotnetHelper);
        }

        public static ValueTask<bool> Beep(int vol, int freq, int duration)
        {
            //https://odino.org/emit-a-beeping-sound-with-javascript/
            return jsRuntime.InvokeAsync<bool>(
                "LBDJsFunctions.Beep", vol, freq, duration);
        }

        public static ValueTask<bool> PianoPlay(char letter, int vol, int freq)
        {
            //https://odino.org/emit-a-beeping-sound-with-javascript/
            return jsRuntime.InvokeAsync<bool>(
                "LBDJsFunctions.PianoPlay", letter, vol, freq);
        }

        public static ValueTask<bool> PianoStop(char letter)
        {
            //https://odino.org/emit-a-beeping-sound-with-javascript/
            return jsRuntime.InvokeAsync<bool>(
                "LBDJsFunctions.PianoStop", letter);
        }
    }
}

File: LupusaBlazorDemos/wwwroot/LBDJsInterop.js
var audioCtx;
var Osctor;

function getArrayBufferFromFileAsync(file, o) {
    const reader = new FileReader();
    return new Promise((resolve, reject) => {
        reader.onload = function () { resolve(reader.result); };
        reader.onprogress = function (e) {
            window.BJSFDEJsFunctions.InvokeOnProgressAction("loadprogress," + e.loaded + "," + e.total);
        };
        reader.onerror = function (err) { reject(err); };

        switch (o) {
            case 0:
                reader.readAsArrayBuffer(file);
                break;
            case 1:
                reader.readAsText(file);
                break;
            default:
                reader.readAsArrayBuffer(file);
        }


    });
}



function dragstart1(e, id, dotnetHelper) {
    e.dataTransfer.setData('text', id.toString());
    dotnetHelper.invokeMethodAsync('InvokeDragStartFromJS', id);
}


function drop1(e, el, id, dotnetHelper) {

    if (e.preventDefault) { e.preventDefault(); }
    if (e.stopPropagation) { e.stopPropagation(); }

    dotnetHelper.invokeMethodAsync('InvokeDropFromJS', [id, e.dataTransfer.getData("text")]);

    document.getElementById(el).removeEventListener('dragstart', null);

    return false;
}


window.LBDJsFunctions = {
    calcfib: function (num) {
        return fib(num);
    },
    generateNewUser: function () {
        return faker.name.findName() + ", " + faker.address.country() + ", Age " + (Math.floor(Math.random() * 50) + 15);
    },
    Alert: function (message) {
        return alert(message);
    },
    SetData: function (v, t) {
        window[v] = t;
        return true;
    },
    GetData: function (v) {
        var result = window[v];
        delete window[v];
        return result;
    },
    ProcessData: function (v) {
        window[v] = window[v].split("").reverse().join("");
    },
    HasFile: async function (inputFile) {
        return document.getElementById(inputFile).files.length > 0;
    },
    ReadFile: async function (v, inputFile) {
        window[v] = await getArrayBufferFromFileAsync(document.getElementById(inputFile).files[0], 0);
        window.BJSFDEJsFunctions.InvokeOnMessageAction("fileloadingdone");
        return true;
    },
    GetFile: async function (v, inputFile) {
        return await getArrayBufferFromFileAsync(document.getElementById(inputFile).files[0], 1);
    },
    handleDragStart: function (el, id, dotnetHelper) {
        if (document.getElementById(el) !== null) {

            document.getElementById(el).addEventListener('dragstart', function (e) { dragstart1(e, id, dotnetHelper); }, true);
            return true;
        }
        else {
            console.log("can't find element");
        }

        return false;
    },
    handleDrop: function (el, id, dotnetHelper) {
        if (document.getElementById(el) !== null) {
            document.getElementById(el).addEventListener('drop', function (e) { drop1(e, el, id, dotnetHelper); }, true);
            return true;
        }
        else {
            console.log("can't find element");
        }

        return false;
    },
    Beep: function (vol, freq, duration) {
        if (audioCtx === undefined) {
            audioCtx = new AudioContext();
        }
        v = audioCtx.createOscillator();
        u = audioCtx.createGain();
        v.connect(u);
        v.frequency.value = freq;
        v.type = "square";
        u.connect(audioCtx.destination);
        u.gain.value = vol * 0.01;
        v.start(audioCtx.currentTime);
        v.stop(audioCtx.currentTime + duration * 0.001);

        return true;
    },
    PianoPlay: function (vol, freq) {
        if (audioCtx === undefined) {
            audioCtx = new AudioContext();
        }

        
        Osctor = audioCtx.createOscillator();
        u = audioCtx.createGain();
        Osctor.connect(u);
        Osctor.frequency.value = freq;
        Osctor.type = "square";
        u.connect(audioCtx.destination);
        u.gain.value = vol * 0.01;
        Osctor.start(audioCtx.currentTime);
        //v.stop(audioCtx.currentTime + duration * 0.001);

        return true;
    },
    PianoStop: function () {
        if (Osctor !== undefined) {
            Osctor.stop();
        }
        return true;
    }
};